### ISSUE TITLE
Significant precision lost with getModelDump() is affecting the model's performance when exported

### ISSUE THREAD

**Issue by gligorijevic** [dmlc/xgboost#7786]
Dear all,

My team has a need to export XGBoost trained model into a c++ code for integration. We were able to export the model using the trees split information in the getModelDump() output, however, the output of the getModelDump() only provides *9 decimals precision* as per an [older pull request](https://github.com/dmlc/xgboost/pull/3298/commits/c3928b3a17b3ccaf9b9125398a2d78cef837091f#), while the XGBoost seems to run on full double precision with *14 decimals*.

Exporting the model we are able to mainly reconstruct the prediction values, however there two big problems when using models with such precision loss for some examples:
- The cut precision can lead to completely wrong decision splits
- For binary classification tasks, summing all leaf values and passing them through a Sigmoid function will yield different result at 3rd plus decimal.

This leads to significant drop of performance as compared to using XGBoost library predictions itself (i.e. ROC AUC would drop by 0.5% to 2% depending on the dataset).

We experimented with training XGBoost on GPU machines that work with half precision and, while there is still some impact, the drop in performance is negligible. 

Thus my question is twofold:
1. What is an alternative way to iterate through decision trees and split criteria for XGBoost model to getModelDump() output?
2. What was the reason to go for `bst_float` precision `int floatMaxPrecision = std::numeric_limits<bst_float>::max_digits10;` instead of full `double` precision like so: `int doubleMaxPrecision = std::numeric_limits<double>::max_digits10;` in `src/tree/tree_model.cc`? Could someone share some insight if making this change into PR would cause any problem?

Thank you!

**Comment by hcho3** [comment]
Also keep in mind that XGBoost uses 32-bit floats to store thresholds.

**Comment by gligorijevic** [comment]
@trivialfis Thanks a lot for the reply! I have looked into the .json output of saveModel() function but there are a few limitations that are preventing me from using it:
- The split criteria has 1 decimal point less than getModelDump() which would further increase issues in reproduceability
- I couldn't find which key refers to features
- I couldn't find which key refers to leaf predictions, or even find predictions as from the output of getModelDump()

Overall I wasn't lucky in finding good interpretation of saveModel() function output. 

@hcho3 Are you positive that Scala XGBoost implementation uses 32-bit floats internally? 

Unless there is some other terms to be added to cumulative score passed to Sigmoid function (reference of which I was not able to find either) - precision loss when outputting the model would be the biggest suspect.
I should mention here that we used GBM implementation of SKlearn and R and got similar effect when we would manually "cut" precision of decision splits and leaf scores (and the two packages work internally with 64-bit precision).

**Comment by hcho3** [comment]
Yes, XGBoost consistently uses 32-bit float internally.

**Comment by trivialfis** [comment]
> The split criteria has 1 decimal 

It's not about how many digits, but can we restore the binary representation of float from decimal representation written in string.  XGBoost adopts the ryu algorithm for serializing floating point, for details please see https://github.com/ulfjack/ryu Baseline is, you can restore the floating-point without any loss.

Also, `UBJSON` is a binary representation of JSON, which you can use without any string parsing.  Visit https://ubjson.org/libraries/ for existing library implementations.

> I couldn't find which key refers to features

Does the example linked in previous comment help?

> Are you positive that Scala XGBoost implementation uses 32-bit floats internally?

Yes.

**Comment by gligorijevic** [comment]
@trivialfis and @hcho3 thanks a lot for confirming that Scala XGBoost implementation uses 32-bit floats internally.

However, unfortunately I must follow up with a few more questions:
1. The prediction output of Scala XGBoost has **18 decimals of precision** which means it has 64-bit representation. How would it be possible to obtain 64-bit number by summing n 32-bit numbers and passing them through Sigmoid function?
3. If we assume that prediction for each examples is obtained by:
    - Sum leaf values for every boosting tree
    - Apply Sigmoid function on the sum
    Are we correct in understanding how XGBoost inference works?
    This way is exactly correct for R's and SKLearn's GBM models; For XGBoost this returns the correct output numbers to the 7th decimal for most examples, and for the rest it starts making errors going as far as the second decimal, which is a huge issue!
    Some [papers (see Fig. 10)](https://drive.google.com/file/d/0B0c0MbnP6Nn-eUNRRkVOOGpkbFk/view?resourcekey=0-nVw3WhovKW5FPvPM5GPHfg) suggest that this is the correct way to do it, [Other resources](https://www.youtube.com/watch?v=8b1JEDvenQU) suggest that we should multiply each leaf with _eta_ which returns completely different results.
    Out hypothesis was that there must be precision lost when exporting the XGBoost model which would cause differences in Sigmoid output, but if this is not the case - I would assume that producing same outputs as XGBoost requires something else to be done in inference stage different from [Fig. 10 here](https://drive.google.com/file/d/0B0c0MbnP6Nn-eUNRRkVOOGpkbFk/view?resourcekey=0-nVw3WhovKW5FPvPM5GPHfg)?

**Comment by trivialfis** [comment]
Hi, the dump model is only used for human inspection, please use save model for precise model persistent.  The latter can be loaded back to XGBoost while the former cannot.


> Are we correct in understanding how XGBoost inference works?

Yes.

**Comment by gligorijevic** [comment]
Hi there so I have looked models outputted into .json file, but given that precision is further cut even to getModelDump() I am not sure that it can reproduce output of XGBoost library.

XGBoost library definitely produces 64-bit output which is making a significant impact on performance on learned ensemble; coming back to my original question. I agree that the `tree_model` ([code](https://github.com/dmlc/xgboost/blob/master/src/tree/tree_model.cc)) is really using `bst_float` and not double, but what is very confusing now is how is XGBoost library in both Python and Scala getting 64-bit outputs?
And more importantly, can how can we **exactly** replicate outputs of XGBoost model using the learned trees?

**Comment by trivialfis** [comment]
One can get back the exact floating point from JSON model, which is mathematically proven by ryu and test proven by enumerating ALL 32-bit floating point numbers (the test is currently in XGB's c++ test suit).  For details about the prove, I recommend reading their paper.

> can how can we exactly replicate

Use XGBoost.  There's no other way for bit by bit replication, especially in parallel computing environment.  This is caused by the non-associative natural of floating point addition.

**Comment by trivialfis** [comment]
Closing as this can be resolved by using either JSON or UBJ.

**Comment by gligorijevic** [comment]
Thank you for your replies @trivialfis, they helped us understand the problems much better!

We made an effort to analyze deeper the source code and see if we would be able to replicate scores of XGBoost library, though as you suggested it does seem to be a very challenging task. 

Having said that, the precision within 32 bit precision is showing >99.5% overlap with extracted GBM model, though comparing with full double precision the number is more closer to 50%. Hopefully XGBoost will be able to share more how the double precision scores are obtained in the future.
